### DISCUSSION TITLE
Sort both the test class and the test method

### DISCUSSION THREAD

**Discussion by tianya66** [pytest-dev/pytest-order#44] [Q&A]
![image](https://user-images.githubusercontent.com/60034363/120276323-204abb80-c2e5-11eb-803d-c20090b06a07.png)
I have a scene like this，
now：
test_a/test_a1.py::Testa01::test_03 
test_a/test_a1.py::Testa01::test_02 
test_a/test_a1.py::Testa01::test_01 

test_a/test_a2.py::Testa02::test_03 
test_a/test_a2.py::Testa02::test_02 
test_a/test_a2.py::Testa02::test_01 

test_b/test_b1.py::Testb01::test_03 
test_b/test_b1.py::Testb01::test_02 
test_b/test_b1.py::Testb01::test_01 

test_b/test_b2.py::Testb02::test_03 
test_b/test_b2.py::Testb02::test_02 
test_b/test_b2.py::Testb02::test_01 
except：
test_a/test_a1.py::Testa01::test_03 
test_a/test_a1.py::Testa01::test_02 
test_a/test_a1.py::Testa01::test_01 

test_b/test_b1.py::Testb01::test_03 
test_b/test_b1.py::Testb01::test_02 
test_b/test_b1.py::Testb01::test_01

test_a/test_a2.py::Testa02::test_03 
test_a/test_a2.py::Testa02::test_02 
test_a/test_a2.py::Testa02::test_01 

test_b/test_b2.py::Testb02::test_03 
test_b/test_b2.py::Testb02::test_02 
test_b/test_b2.py::Testb02::test_01

**Comment by mrbean-bremen** [comment]
That is actually an interesting case.
In order to sort the classes, but not the tests inside the classes you need `--order-group-scope=class`. If you use `--order-scope=class`, sorting would only be done inside the classes.
To sort the classes, you can use relative sorting. In your case, this has to be done on the test level because `pytest-order` does not support separate markers for the same test (e.g. one marker at class level, and one at test level), which I hadn't noticed before - I just wrote an issue for that.
In your example, it should be sufficient to make one of the tests in `Testa01` dependent on a test in `Testb02`:
```py

class Testa02:
    @pytest.mark.order(
        index=3,
        after="test_b/test_b1.py::Testb01::test_01"
    )
    def test_01(self):
        pass

    @pytest.mark.order(2)
    def test_02(self):
        pass

    @pytest.mark.order(1)
    def test_03(self):
        pass
````
This ensures that `Testa01` is tested after `Testb02`, and the order of the other tests is not changed, e.g. they will be executed in the order:
`Testa01` -> `Testb01` -> `Testa02` -> `Testb02`

A few notes:
- it is important (at the moment) to write both the `index` and the `after` attributes into the same marker, as more than one marker for the same test is not supported (I created #45 for this)
- it does not matter which of the tests in the class you make dependent on a test in the other class, because the order inside the class will not change with the `--order-group-scope=class` option
- the syntax for referencing a test in another module has slightly changed in version 1.0.0, so this has to be changed a bit if you use a pre-1.0 version of `pytest-order`

**Reply by mrbean-bremen** [reply]
After #45 is addressed, you will be able to use relative markers at class level, which shall be more convenient.

**Reply by tianya66** [reply]
Thanks for your reply。

**Reply by mrbean-bremen** [reply]
@tianya66 - did that work for you?

**Reply by mrbean-bremen** [reply]
With current main, this should also work:
```py
@pytest.mark.order(after="test_b/test_b1.py::Testb01")
class Testa02:
    @pytest.mark.order(3)
    def test_01(self):
        pass

    @pytest.mark.order(2)
    def test_02(self):
        pass

    @pytest.mark.order(1)
    def test_03(self):
        pass
```

**Reply by tianya66** [reply]
ooooooook!!!!!!!!! Thank you for your answer。

**Comment by tianya66** [comment]
sorry ，I use pip install pytest-order,  this version cant work it.   virsion=1.0.0



***@***.***
 
From: mrbean-bremen
Date: 2021-06-08 03:04
To: pytest-dev/pytest-order
CC: tianya; Mention
Subject: Re: [pytest-dev/pytest-order] Sort both the test class and the test method (#44)
With current master, this should also work:
@pytest.mark.order(after="test_b/test_b1.py::Testb01")
class Testa02:
    @pytest.mark.order(3)
    def test_01(self):
        pass
    @pytest.mark.order(2)
    def test_02(self):
        pass
    @pytest.mark.order(1)
    def test_03(self):
        pass

**Reply by mrbean-bremen** [reply]
As I wrote, this is in main (e.g. not yet released), you can [install](https://pytest-dev.github.io/pytest-order/stable/intro.html#installation) it using:
```
pip install git+https://github.com/pytest-dev/pytest-order
```
EDIT: meanwhile it also in release mode, so it can be installed vie PyPi as usual
